### PR TITLE
feat: Ability to load the Caseworker and IDAM-Caseworker Role Mapping files using the Bulk Upload API

### PR DESCRIPTION
**Pull request by SujeethKumarAri**
### JIRA link  ###

https://tools.hmcts.net/jira/browse/RDCC-2216

### Change description ###

- Ability to load the Caseworker and IDAM-Caseworker Role Mapping files using the Bulk Upload API
- Have added code to handle cell type Formula for functional test case for now. But this code will be removed before deploying the application to prod
-  Have added loggers to log time taken for few methods to be processed. But this code will be removed before deploying the application to prod.
- I have created https://tools.hmcts.net/jira/browse/RDCC-2262 to remove the non-production related codes

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    @Autowired
+    CaseWorkerProfileConverter caseWorkerProfileConverter;
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public ResponseEntity<Object> processFile(MultipartFile file) {
+        long time1 = System.currentTimeMillis();
+        Workbook workbook = excelValidatorService.validateExcelFile(file);
+        log.info("----Time taken to validate the given file "
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java:46]: you can print file name in loggers of time taken

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                + (System.currentTimeMillis() - time2));
+        long time3 = System.currentTimeMillis();
+        List<CaseWorkerDomain> invalidRecords = validationService.getInvalidRecords(caseWorkerRequest);
+        log.info("----Time taken to validate the records "
+                + (System.currentTimeMillis() - time3));
+        if (!invalidRecords.isEmpty()) {
+            caseWorkerRequest.removeAll(invalidRecords);
+        }
+
+        if (caseWorkerRequest.get(0).getClass().isAssignableFrom(CaseWorkerProfile.class)) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java:65]: proceed if at least 1 valid records. i.e. check if empty

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java (hunk 1 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java:46]: sure

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java (hunk 2 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceFacadeImpl.java:65]: sure

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/client/domain/ServiceRoleMapping.java [hunk 3]
```diff
@@ -19,7 +24,7 @@
     String serivceId;
 
     @MappingField(columnName = "Roles")
-    @NotEmpty
+    @NotNull
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/client/domain/ServiceRoleMapping.java:27]: I think we need change primitive to Integer wrapper else @Notnull never catch up.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+     *
+     * @param caseWorkerProfiles the source object to convert,
+     *                           which must be an instance of {@code CaseWorkerDomain} (never {@code null})
+     * @return the converted object, which must be an instance of {@code CaseWorkersProfileCreationRequest}
+     *              (potentially {@code null})
+     * @throws IllegalArgumentException if the source cannot be converted to the desired target type
+     */
+    public List<CaseWorkersProfileCreationRequest> convert(List<CaseWorkerDomain> caseWorkerProfiles) {
+        List<CaseWorkersProfileCreationRequest> caseWorkersProfileCreationRequests = new ArrayList<>();
+        caseWorkerProfiles
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java:39]: ofNullable().ifPresent may validate else null pointer may get

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java [hunk 5]
```diff
@@ -32,16 +33,25 @@
...
             throw new ExcelValidationException(HttpStatus.BAD_REQUEST, FILE_NO_DATA_ERROR_MESSAGE);
         }
-        Sheet sheet = workbook.getSheet(CaseWorkerConstants.REQUIRED_SHEET_NAME);
+        Sheet sheet;
+        if (classType.isAssignableFrom(CaseWorkerProfile.class)) {
+            sheet = workbook.getSheet(REQUIRED_CW_SHEET_NAME);
+        } else {
+            sheet = workbook.getSheet(REQUIRED_ROLE_MAPPING_SHEET_NAME);
+        }
         if (null == sheet) {
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java:55]: Objects.IsNull you can use

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java (hunk 4 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java:39]: we are checking whether the list is empty or not before calling convert in my latest change

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java (hunk 5 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java:55]: if we go inside the IsNull it is doing the same
`public static boolean isNull(Object obj) {
        return obj == null;
    }`
    any reason to use it ?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefController.java [hunk 6]
```diff
@@ -90,16 +86,11 @@
...
-                .message(CaseWorkerConstants.REQUEST_COMPLETED_SUCCESSFULLY)
-                .build();
-
-        return ResponseEntity
-                .status(201)
-                .body(uploadCaseWorkerFileResponse);
+        long time1 = System.currentTimeMillis();
+        ResponseEntity<Object> responseEntity = caseWorkerServiceFacade.processFile(file);
+        log.info("----Time taken to upload the given file "
+                + (System.currentTimeMillis() - time1));
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefController.java:92]: {} use placeholders rather + and also please use logging component name to track this in app insights

Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/CaseWorkerRefController.java:92]: sure

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/internal/impl/CaseWorkerInternalApiClientImpl.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import javax.servlet.http.HttpServletRequest;
+
+import static java.util.Objects.nonNull;
+import static org.springframework.http.MediaType.APPLICATION_JSON;
+
+@Service
+public class CaseWorkerInternalApiClientImpl implements CaseWorkerInternalApiClient {
+
+    public static final String SERVICE_AUTHORIZATION = "ServiceAuthorization";
+    public static final String AUTHORIZATION = "Authorization";
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/internal/impl/CaseWorkerInternalApiClientImpl.java:26]: move these constants in util or constant class may be we can re use in other places if required.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                            CaseWorkersProfileCreationRequest
+                            .caseWorkersProfileCreationRequest()
+                            .firstName(obj.getLastName())
+                            .lastName(obj.getLastName())
+                            .emailId(obj.getOfficialEmail())
+                            .region(obj.getRegionName())
+                            .regionId(obj.getRegionId())
+                            .suspended(Boolean.parseBoolean(obj.getSuspended()))
+                            .userType(obj.getUserType())
+                            .idamRoles(null == obj.getIdamRoles() ? null :
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java:53]: try avoid null where ever possible isNull we can use

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/internal/impl/CaseWorkerInternalApiClientImpl.java (hunk 7 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/internal/impl/CaseWorkerInternalApiClientImpl.java:26]: sure

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java (hunk 8 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java:53]: if we go inside the IsNull it is doing the same
public static boolean isNull(Object obj) { return obj == null; }
any reason to use it ?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java (hunk 5 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java:55]: it is using things aviable rather we writing.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java (hunk 4 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/CaseWorkerProfileConverter.java:39]: ok

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java (hunk 5 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/ExcelAdaptorServiceImpl.java:55]: ok then.. I shall use Objects.IsNull

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/client/domain/ServiceRoleMapping.java (hunk 3 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/client/domain/ServiceRoleMapping.java:27]: Sure

**Review by shreedhar-hmcts — DISMISSED** [review]

**Review by shreedhar-hmcts — APPROVED** [review]

**Review by nitishsingla1234 — DISMISSED** [review]

**Review by shreedhar-hmcts — APPROVED** [review]

**Review by abhijitdiwan — APPROVED** [review]

**Pull request merged**